Add two related tests about economic viability

harisang opened the pull request:
This PR adds a test around economic viability of the protocol. Specifically, the following are checked:

1. Fees collected - execution cost, per settlement. Results are aggregated per solver and reported once a day.
2. Fees collected - payout to solver, per settlement. Again, results are aggregated per solver and reported once a day.

fhenneke commented on src/monitoring_tests/cost_coverage_per_solver_test.py:
```diff
@@ -0,0 +1,104 @@
+        solver = solution["solver"]
+        ref_score = 0
+        if len(competition_data["solutions"]) > 1:
+            second_best_sol = competition_data["solutions"][-1]
```
the second largest score should come from this solution:
```suggestion
            second_best_sol = competition_data["solutions"][-2]
```
(might not exist though)

harisang replied:
Oh good catch, this was a typo. Since we check the number of solutions, then we are sure it exists at this point of the code.

fhenneke replied:
Ok, the change looks good then.

fhenneke commented on src/monitoring_tests/cost_coverage_per_solver_test.py:
```diff
@@ -0,0 +1,104 @@
+                ref_score = second_best_sol["scoreProtocol"]
+        if solver in self.cost_coverage_per_solver:
+            self.cost_coverage_per_solver[solver] += fees - gas_cost
+            self.total_coverage_per_solver[solver] += ref_score - surplus
```
This looks like `fee - payment` (since `fee - payment = fee - (surplus + fee - ref_score) = ref_score - surplus`. Above you wrote it is `payment - cost`.
Also, it does not take into account capping of payments. Though we might not do that on purpose here. (In some scripts, e.g. the solver competition monitoring, I have both a capped and an uncapped query. But this is always a bit messy.)

harisang replied:
Right, so the comment at the top is wrong, my intention to really look at what the protocol collects and compare it with the payout to the solver.

And you are right that I completely missed the capping. Thanks for catching this, I will fix this, as I would like to keep track of what is actually going on now.

harisang replied:
Updated this so would be good to check if you agree now with it.

fhenneke replied:
Sound good.

fhenneke commented on src/monitoring_tests/cost_coverage_per_solver_test.py:
```diff
@@ -0,0 +1,104 @@
+
+        ### This part takes care of the reporting once a day.
+        current_block = self.web3_api.get_current_block_number()
+        if current_block is None or self.original_block is None:
```
Since `self.original_block` can be `None` from initialization, this might always fail. Some mechanism for setting `self.original_block` if the initial call fails would be required.

harisang replied:
Very good catch!

I added an additional step here. Let me know if that makes sense.

fhenneke replied:
Looks good to me.

fhenneke commented on src/monitoring_tests/cost_coverage_per_solver_test.py:
```diff
@@ -0,0 +1,104 @@
+        current_block = self.web3_api.get_current_block_number()
+        if current_block is None or self.original_block is None:
+            return success
+        if current_block - self.original_block > DAY_BLOCK_INTERVAL:
```
We could also loop through all solvers here to get a daily overview over everything.
Otherwise we might get that report with widely different time horizons for less successful solvers.

fhenneke reviewed (changes requested):
I added some minor comments in the code.

If I understood correctly, this test aggregates data over the day and then creates an info log message.

Is the goal to also trigger alerts for unusually large or small values? Or generate alerts on a per auction basis?

Do you think adding this test to the EBBO repo is more reliable or easier to maintain compared to a Dune dashboard? All the data we use in this PR would also available there.

harisang commented:
> I added some minor comments in the code.
> 
> If I understood correctly, this test aggregates data over the day and then creates an info log message.
> 
> Is the goal to also trigger alerts for unusually large or small values? Or generate alerts on a per auction basis?
> 
> Do you think adding this test to the EBBO repo is more reliable or easier to maintain compared to a Dune dashboard? All the data we use in this PR would also available there.

Good point. Tbh, it is mostly my insecurity with Dune, when and how is the data fetched there, and how easily is to maintain queries there etc.

I don't want to push for breaking away from Dune, but i just feel it is more versatile and flexible to work directly with our data, and it is easier to add other things to this, generate alerts etc. But this is more of a personal opinion rather than robust arguments.

fhenneke commented:
Did the test run locally for you (with reduced `DAY_BLOCK_INTERVAL`)?
I changed a few lines with regads to `ref_score` (see comment) and then it worked fine.

harisang commented:
> Did the test run locally for you (with reduced `DAY_BLOCK_INTERVAL`)? I changed a few lines with regads to `ref_score` (see comment) and then it worked fine.

Didn't actually see any comment of yours, but your hint about the `ref_score` was enough. Fixed it (maybe not in the nicest way) but it should work now.

fhenneke commented on src/monitoring_tests/cost_coverage_per_solver_test.py:
```diff
@@ -0,0 +1,108 @@
+                ref_score = second_best_sol["scoreDiscounted"]
+            else:
+                ref_score = second_best_sol["scoreProtocol"]
+        payout = surplus + fees - ref_score
```
Fixed

fhenneke commented:
Oh, I just added comments to a "review", but did not submit the review. Thus all comments were not actually posted.
Before I always just added single comments instead of comments of a review.
